python/lsst/daf/butler/_butler.py

```diff
@@ -799,6 +799,23 @@ def get_dataset_type(self, name: str) -> DatasetType:
         """
         raise NotImplementedError()
 
+    @abstractmethod
+    def get_dataset(self, id: DatasetId) -> DatasetRef | None:
```

Comment:
There are two open questions:

* If we have vectorized version, `get_datasets` do we keep `get_dataset` around?
* What should `get_datasets` return? `list[DatasetRef]` or `DatasetRefContainer`?

I think we likely want a single request to return a single ref so it seems like we always want `Butler.get_dataset` even if we later add `Butler.get_datasets`.

Reply:
It certainly makes client/server testing a lot easier to begin with since it's the simplest possible API to implement. We do have VO services that are given a UUID (via datalink from, eg, ObsTAP) and need to convert that to a dataset ref so it's not impossible that the server will need to do that (depending on how services connect to butlers).

Reply:
I think we'll want to start exposing dataset IDs in more places, too; I've been thinking about how we might to get them into the `--where` expression language, for instance. They appear pretty prominently in QGs already.